# 01 – Details

Esplorazione e data cleaning del dataset `details.csv`.

**Colonne:**
| Colonna | Descrizione |
|---|---|
| `mal_id` | ID univoco dell'anime su MyAnimeList |
| `title` | Titolo dell'anime |
| `title_japanese` | Titolo in giapponese |
| `url` | URL della pagina MAL dell'anime |
| `image_url` | URL dell'immagine di copertina |
| `type` | Tipo (TV, Movie, etc) |
| `status` | Stato di trasmissione |
| `score` | Punteggio medio |
| `scored_by` | Numero di utenti che hanno votato |
| `start_date` | Data di inizio trasmissione |
| `end_date` | Data di fine trasmissione |
| `synopsis` | Sinossi testuale |
| `rank` | Classifica globale |
| `popularity` | Popolarità |
| `members` | Numero di utenti che hanno l'anime nella loro lista |
| `favorites` | Numero di utenti che hanno messo l'anime tra i preferiti |
| `genres` | Lista di generi |
| `studios` | Lista di studi di produzione |
| `themes` | Lista di temi narrativi |
| `demographics` | Lista di target demografici |
| `source` | Materiale di origine (Manga, Novel, Original, etc) |
| `rating` | Classificazione per età |
| `episodes` | Numero di episodi |
| `season` | Stagione di uscita |
| `year` | Anno di uscita |
| `producers` | Lista di produttori |
| `explicit_genres` | Lista di generi espliciti |
| `licensors` | Lista di licenziatari |
| `streaming` | Lista di piattaforme di streaming |

## 1. Import e caricamento dati
Importiamo le librerie necessarie e carichiamo il file csv. Facciamo una esplorazione generica per capire la struttura e le caratteristiche del dataset.

In [39]:
import pandas as pd
import numpy as np
from dataset_analyzer import analyze

df_details = pd.read_csv('../datasets/details.csv')
print(f'Shape: {df_details.shape}')
print()
df_details.info()
df_details.head()

Shape: (28955, 29)

<class 'pandas.DataFrame'>
RangeIndex: 28955 entries, 0 to 28954
Data columns (total 29 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   mal_id           28955 non-null  int64  
 1   title            28955 non-null  str    
 2   title_japanese   28836 non-null  str    
 3   url              28955 non-null  str    
 4   image_url        28955 non-null  str    
 5   type             28888 non-null  str    
 6   status           28955 non-null  str    
 7   score            18882 non-null  float64
 8   scored_by        18882 non-null  float64
 9   start_date       28104 non-null  str    
 10  end_date         11267 non-null  str    
 11  synopsis         23828 non-null  str    
 12  rank             21997 non-null  float64
 13  popularity       28955 non-null  int64  
 14  members          28955 non-null  int64  
 15  favorites        28955 non-null  int64  
 16  genres           28955 non-null  str    
 17  stu

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,demographics,source,rating,episodes,season,year,producers,explicit_genres,licensors,streaming
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,NaN,2010-01-01T00:00:00+00:00,...,[],Original,G - All Ages,1.0,NaN,NaN,['Nagoya Zokei University'],[],[],[]
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503.0,2023-06-11T00:00:00+00:00,...,[],Original,PG-13 - Teens 13 or older,1.0,NaN,NaN,[],[],[],[]
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745.0,2007-01-18T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,"['Bandai Visual', 'CyberConnect2']",[],[],[]
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373.0,2007-12-22T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],"['Funimation', 'Bandai Entertainment']",[]
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317.0,2008-03-25T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],[],[]


Il dataset contiene **28.955 righe** e **29 colonne**. Alcune colonne presentano valori nulli. Alcuni dtype richiedono conversione: le date sono in formato stringa, e alcune colonne intere (`scored_by`, `rank`, `episodes`, `year`) sono memorizzate come `float64`. Quiesti veranno analizzati e gestiti singolarmente nella sezione 2.

## 1.1 Rimozione duplicati esatti

Prima dell'analisi per colonna, rimuoviamo le righe con valori identici in **tutte** le colonne, mantenendo solo la prima occorrenza.

In [40]:
n_originale = len(df_details)

mask_dup = df_details.duplicated(keep=False)
n_righe_coinvolte = mask_dup.sum()
n_gruppi = df_details[mask_dup].duplicated(keep='first').sum()
n_tenute = n_righe_coinvolte - n_gruppi

print(f'Righe totali coinvolte in duplicazioni : {n_righe_coinvolte:,}')
print(f'  → prime occorrenze mantenute         : {n_tenute:,}')
print(f'  → occorrenze extra rimosse           : {n_gruppi:,}')
print()

df_details.drop_duplicates(keep='first', inplace=True)
print(f'Righe prima della rimozione : {n_originale:,}')
print(f'Righe dopo la rimozione     : {len(df_details):,}')

Righe totali coinvolte in duplicazioni : 0
  → prime occorrenze mantenute         : 0
  → occorrenze extra rimosse           : 0

Righe prima della rimozione : 28,955
Righe dopo la rimozione     : 28,955


## 2. Analisi colonna per colonna

### 2.1 `mal_id`

È la **chiave primaria** del dataset. Deve essere intera, non nulla e univoca. Utilizziamo `analyze` per verificare queste proprietà.

In [41]:
analyze(df_details['mal_id'])


  Nome serie                     mal_id
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   28,955  (100.00%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            62,590
  Range                          62,589
  Media                          33,977.52
  Mediana                        37,969.00
  Moda/e                         1, 5, 6
  Dev. standard    

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un ID univoco.
- **Nessun duplicato**: ogni riga corrisponde a un anime distinto.
- **Dtype corretto**: `int64`, nessuna conversione necessaria.

**Nessuna pulizia necessaria.**

### 2.2 `title`

Colonna testuale con il titolo dell'anime. Verifichiamo la presenza di valori nulli e duplicati.

In [42]:
df_details['title'] = df_details['title'].str.strip()
analyze(df_details['title'])


  Nome serie                     title
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   28,953  (99.99%)
  Valori duplicati               2  (0.01%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  186  caratteri
  Lunghezza media                24.70  caratteri
  Lunghezza mediana              21.0000  caratteri
  Dev. standard lunghezza        15.07
  IQR lunghezza                  18.0000

Valori di lunghez

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un titolo.
- **Quasi univoco**: 28.953 valori distinti su 28.955. Ci solo 2 titoli duplicati che verifichiamo successivamente.
- **Lunghezze variabili**: da 1 carattere a 186 caratteri. Stiampiamo i 10 più lunghi e 10 più corti e facciamo delle verifiche per approfondire.
- **14 titoli solo numerici**: Indaghiamo in seguito.
- **1 titolo contiene un URL**: Ingaghiamo in seguito.
- **43.4% contiene caratteri speciali**: normale, i titoli includono punteggiatura, simboli e trattini.

In [43]:
# Titoli duplicati
dup_mask = df_details.duplicated(subset=['title'], keep=False)
print(f'Titoli duplicati (righe coinvolte): {dup_mask.sum()}')
print()
print(df_details[dup_mask][['mal_id', 'title', 'type', 'status', 'year']].to_string(index=False))

# Titoli solo numerici
print()
print('─' * 60)
num_mask = df_details['title'].str.fullmatch(r'\d+')
print(f'Titoli solo numerici: {num_mask.sum()}')
print()
print(df_details[num_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Titolo che contiene un URL
print()
print('─' * 60)
url_mask = df_details['title'].str.contains(r'https?://', regex=True, na=False)
print(f'Titoli con URL: {url_mask.sum()}')
print()
print(df_details[url_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Lunghezze estreme
print()
print('─' * 60)
lunghezze = df_details['title'].str.len()
top10_lunghi = df_details.loc[lunghezze.nlargest(10).index][['mal_id', 'title', 'type', 'year']]
top10_corti  = df_details.loc[lunghezze.nsmallest(10).index][['mal_id', 'title', 'type', 'year']]
print('10 titoli più lunghi:')
print(top10_lunghi.to_string(index=False))
print()
print('10 titoli più corti:')
print(top10_corti.to_string(index=False))

# Titoli con caratteri speciali
print()
print('─' * 60)
spec_mask = df_details['title'].str.contains(r'[^a-zA-Z0-9\s]', regex=True, na=False)
print(f'Titoli con caratteri speciali: {spec_mask.sum()}')
print()
print('Primi 10:')
print(df_details[spec_mask][['mal_id', 'title', 'type', 'year']].head(10).to_string(index=False))

Titoli duplicati (righe coinvolte): 4

 mal_id                                    title  type          status  year
  59994                                Moonlight Music Finished Airing   NaN
  60524                                Moonlight Music Finished Airing   NaN
  53367 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN
  55740 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN

────────────────────────────────────────────────────────────
Titoli solo numerici: 14

 mal_id  title  type   year
  29978    001   ONA    NaN
  60145   1135 Music    NaN
  37065    123 Music    NaN
  55204   1985 Music    NaN
  30234   1989 Music    NaN
  38756      2 Music    NaN
  54066   2005 Music    NaN
  32813   2010 Music    NaN
  53789     32 Music    NaN
  53771   3341 Music    NaN
  60486    520 Movie    NaN
  13971 663114 Movie    NaN
  41457     86    TV 2021.0
  36089      9 Music    NaN

────────────────────────────────────────────────────────────
Titoli

**Risultati dell'indagine:**

- **Titoli duplicati (2 coppie, 4 righe):** `Moonlight` e `Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian` compaiono ciascuno due volte, ma con `mal_id` distinti. Sono anime separati su MAL che condividono il titolo. **Nessuna rimozione necessaria.**

- **Titoli solo numerici (14 righe):** tutti legittimi. La maggior parte sono video musicali (`Music`) o cortometraggi in cui titoli brevi o numerici sono comuni. **Nessuna azione necessaria.**

- **Titolo contenente un URL (1 riga):** `Himitsukessha Taka no Tsume The Movie 3: http://takanotsume.jp wa Eien ni` (`mal_id` 8357). L'URL è parte integrante del titolo ufficiale del film, caratteristica della serie *Taka no Tsume* nota per titoli volutamente assurdi. **Nessuna azione necessaria.**

- **Lunghezze estreme:** i titoli molto corti (1–2 caratteri) sono anime reali con titoli brevi; i titoli molto lunghi (fino a 186 caratteri) sono tipici degli produzioni moderni. Nessun valore anomalo.

- 43.4% dei titoli contiene caratteri speciali (12.572): percentuale alta ma attesa. Dai primi 10 stampati si nota che si tratta di punteggiatura e simboli che fanno   parte naturale titoli anime.

**Conclusione:** la colonna `title` non richiede alcuna pulizia. Tutti i casi apparentemente anomali sono valori legittimi.

### 2.3 `title_japanese`

Colonna testuale con il titolo in giapponese. I null sono ammessi: anime di produzione non giapponese o senza titolo kanji registrato su MAL non avranno questo campo.

In [44]:
df_details['title_japanese'] = df_details['title_japanese'].str.strip()
analyze(df_details['title_japanese'])


  Nome serie                     title_japanese
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,836  (99.59%)
  Null / NaN                     119  (0.41%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   27,772  (95.91%)
  Valori duplicati               1,064  (3.67%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  85  caratteri
  Lunghezza media                12.24  caratteri
  Lunghezza mediana              10.0000  caratteri
  Dev. standard lunghezza        8.69
  IQR lunghezza                  10.0000

Valor

**Osservazioni:**
- **119 valori nulli** (~0.4%): strutturali, anime di produzione non giapponese o senza titolo kanji registrato su MAL.
- Facciamo un controllo sui titoli duplicati, con lunghezze estreme, contenenti `@` e solo numerici.

In [45]:
cols = ['mal_id', 'title', 'title_japanese', 'type', 'year']

# Duplicati
dup_mask_tj = df_details['title_japanese'].duplicated(keep=False) & df_details['title_japanese'].notna()
print(f'Titoli giapponesi duplicati (righe coinvolte): {dup_mask_tj.sum()}')
print()
print('Primi 10:')
print(df_details[dup_mask_tj][cols].head(10).to_string(index=False))

# Lunghezze estreme
print()
print('─' * 60)
lunghezze_tj = df_details['title_japanese'].str.len()
print('10 titoli giapponesi più lunghi:')
print(df_details.loc[lunghezze_tj.nlargest(10).index, cols].to_string(index=False))
print()
print('10 titoli giapponesi più corti:')
print(df_details.loc[lunghezze_tj.nsmallest(10).index, cols].to_string(index=False))

# Titoli con @
print()
print('─' * 60)
at_mask_tj = df_details['title_japanese'].str.contains('@', regex=False, na=False)
print(f'Titoli giapponesi con @: {at_mask_tj.sum()}')
print()
print(df_details[at_mask_tj][cols].to_string(index=False))

# Solo numerici
print()
print('─' * 60)
num_mask_tj = df_details['title_japanese'].str.fullmatch(r'\d+', na=False)
print(f'Titoli giapponesi solo numerici: {num_mask_tj.sum()}')
print()
print(df_details[num_mask_tj][cols].to_string(index=False))

Titoli giapponesi duplicati (righe coinvolte): 2003

Primi 10:
 mal_id                                                 title      title_japanese       type   year
   3269                                   .hack//G.U. Trilogy .hack//G.U. Trilogy      Movie    NaN
   4469                      .hack//G.U. Trilogy: Parody Mode .hack//G.U. Trilogy    Special    NaN
  41380            100-man no Inochi no Ue ni Ore wa Tatteiru    100万の命の上に俺は立っている         TV 2020.0
  44881 100-man no Inochi no Ue ni Ore wa Tatteiru 2nd Season    100万の命の上に俺は立っている         TV 2021.0
  48976      100-man no Inochi no Ue ni Ore wa Tatteiru Recap    100万の命の上に俺は立っている TV Special    NaN
  32601                   12-sai.: Chicchana Mune no Tokimeki  12歳。～ちっちゃなムネのトキメキ～         TV 2016.0
  33419        12-sai.: Chicchana Mune no Tokimeki 2nd Season  12歳。～ちっちゃなムネのトキメキ～         TV 2016.0
   9624                                30-sai no Hoken Taiiku            30歳の保健体育         TV 2011.0
  12131                       30-sai 

- I titoli giaponesi contenenti `@` corrispondono con titoli in inglese con `@` nella stessa posizione. Il carattere fa parte del titolo. Nessuna pulizia necessaria.
- I titoli giaponesi solo numerici corrispondono a titoli in inglese anche loro solo numerici. Nessuna pulizia necessaria.
- I titoli più corti come osservato dai primi 10 stampati mostrano una corrispondenza con i titoli in inglese e appartengono principalmente a brani musicali o film. Anche questo è atteso.
- I titoli più lunghi seguono la stessa logica dei titoli più corti con l'eccezione di un titolo che è uguale a quello in inglese. Facciamo un controllo per vedere se è un caso isolato o se accade più volte.

In [46]:
# title == title_japanese
same_mask = df_details['title'] == df_details['title_japanese']
print(f'Titoli identici in title e title_japanese: {same_mask.sum()} ({same_mask.mean()*100:.1f}%)')
print()
print('Primi 10:')
print(df_details[same_mask][['mal_id', 'title', 'title_japanese', 'type', 'year']].head(10).to_string(index=False))

Titoli identici in title e title_japanese: 833 (2.9%)

Primi 10:
 mal_id                 title        title_japanese    type   year
  56036                ......                ......   Music    NaN
   3269   .hack//G.U. Trilogy   .hack//G.U. Trilogy   Movie    NaN
   1143     .hack//Intermezzo     .hack//Intermezzo Special    NaN
   9332        .hack//Quantum        .hack//Quantum     OVA    NaN
    873          .hack//Roots          .hack//Roots      TV 2006.0
   1487         .hack//Unison         .hack//Unison Special    NaN
  29978                   001                   001     ONA    NaN
  28761                 00:08                 00:08   Movie    NaN
   7374                  100%                  100%     OVA    NaN
  39208 100% Renewable Energy 100% Renewable Energy Special    NaN


Notiamo che ci sono 833 righe con nome in inglese uguale a quello in giaponese. Potrebbe trattarsi di opere senza nome giaponese. Abbiamo deciso di tenerlo invariato.

**Nessuna pulizia necessaria**

### 2.4 `url`

URL della pagina MAL dell'anime. Verifichiamo la presenza di null, l'unicità, che il format sia corretto e che ci sia coerenza tra `mal_id` e `url`.

In [47]:
df_details['url'] = df_details['url'].str.strip()
analyze(df_details['url'])

pattern_url = r'^https://myanimelist\.net/anime/\d+/\S+$'
non_conformi_url = df_details[~df_details['url'].str.match(pattern_url, na=False)]
print(f'URL non conformi al formato atteso: {len(non_conformi_url)}')

# Consistenza mal_id <-> url
id_from_url = df_details['url'].str.extract(r'/anime/(\d+)/')[0].astype(int)
mismatch_url = id_from_url != df_details['mal_id']
print(f'mal_id non coerente con url        : {mismatch_url.sum()}')


  Nome serie                     url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   28,955  (100.00%)
  Valori duplicati               0  (0.00%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  35  caratteri
  Lunghezza max                  216  caratteri
  Lunghezza media                60.19  caratteri
  Lunghezza mediana              57.0000  caratteri
  Dev. standard lunghezza        14.92
  IQR lunghezza                  17.0000

Valori di lunghez

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un URL valido.
- **Nessun duplicato**: coerente con `mal_id` come chiave primaria.
- Tutti i 28.955 URL seguono il pattern `https://myanimelist.net/anime/{id}/...`. Nessuna anomalia.
- L'ID estratto dall'URL corrisponde esattamente a `mal_id` per ogni riga. Nessuna inconsistenza.

**Nessuna pulizia necessaria.**

### 2.5 `image_url`

URL dell'immagine di copertina. Verifichiamo anche il format del URL.

In [48]:
df_details['image_url'] = df_details['image_url'].str.strip()
analyze(df_details['image_url'])

pattern_img = r'^https://cdn\.myanimelist\.net/(images/anime/\d+/\d+\.(jpg|webp|png)|img/sp/icon/apple-touch-icon-256\.png)$'
non_conformi_img = df_details[~df_details['image_url'].str.match(pattern_img, na=False)]
print(f'image_url non conformi al formato atteso: {len(non_conformi_img)}')


  Nome serie                     image_url
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   28,758  (99.32%)
  Valori duplicati               197  (0.68%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  50  caratteri
  Lunghezza max                  64  caratteri
  Lunghezza media                54.69  caratteri
  Lunghezza mediana              56.0000  caratteri
  Dev. standard lunghezza        1.93
  IQR lunghezza                  3.0000

Valori di lun

**Osservazione**
- Non ci sono stringhe vuote o null. Tutta la colonna è popolata.
- Tutti i 28.955 URL seguono il pattern corretto
- Ci sono 197 URL duplicati che corrispondono a un link che porta a un'immagine placeholder. A ogni personaggio senza immagine reale viene assegnato il logo MAL. Non è un errore e non è necessaria nessuna pulizia.
- Tutte le URL seguono il formato corretto senza nessuna anomalia strutturale.

**Nessuna pulizia necessaria**

### 2.6 `type`

Colonna categorica con il tipo di anime (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special).

In [49]:
df_details['type'] = df_details['type'].str.strip()
analyze(df_details['type'])


  Nome serie                     type
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,888  (99.77%)
  Null / NaN                     67  (0.23%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   9  (0.03%)
  Valori duplicati               28,879  (99.74%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  2  caratteri
  Lunghezza max                  10  caratteri
  Lunghezza media                3.73  caratteri
  Lunghezza mediana              3.0000  caratteri
  Dev. standard lunghezza        1.81
  IQR lunghezza                  3.0000

Valori di lunghezza est

**Osservazioni:**
- **67 valori nulli**: attesi, anime ancora non publicate per cui il tipo non è ancora registrato su MAL.
- **9 valori distinti**: tutti coerenti con la terminologia MAL (TV, Movie, OVA, ONA, Special, Music, CM, PV, TV Special).

**Nessuna pulizia necessaria.**

### 2.7 `status`

Colonna categorica con lo stato di trasmissione.

In [50]:
df_details['status'] = df_details['status'].str.strip()
analyze(df_details['status'])


  Nome serie                     status
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   3  (0.01%)
  Valori duplicati               28,952  (99.99%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  13  caratteri
  Lunghezza max                  16  caratteri
  Lunghezza media                14.97  caratteri
  Lunghezza mediana              15.0000  caratteri
  Dev. standard lunghezza        0.29
  IQR lunghezza                  0.0000

Valori di lunghezz

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno uno stato definito.
- **3 valori distinti**: `Finished Airing`, `Currently Airing`, `Not yet aired`, coerenti con la terminologia MAL.

**Nessuna pulizia necessaria.**

### 2.8 `score`

Punteggio medio assegnato dagli utenti MAL.

In [51]:
analyze(df_details['score'])


  Nome serie                     score
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   561  (1.94%)
  Valori float                   18,737

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1.8900
  Max                            9.2900
  Range                          7.4000
  Media                          6.3905
  Mediana                        6.3600
  Moda/e             

**Osservazioni:**
- **10.073 valori nulli** (~34.8%): in quanto la valutazione va da 1 a 10, un valore null rapresenta l'assenza di valutazioni.
- Il dtype `float64` corretto. Nessuna conversione necessaria.
- Gli outliers non vengono considerati in quanto una valutazione compresa tra 1 e 10 non è un errore.

**Nessuna pulizia necessaria**.

### 2.9 `scored_by`

Numero di utenti che hanno votato.

In [52]:
analyze(df_details['scored_by'])


  Nome serie                     scored_by
  dtype                          float64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               18,882  (65.21%)
  Null / NaN                     10,073  (34.79%)
  Zeri                           0  (0.00%)
  Positivi                       18,882   (65.21%)
  Negativi                       0   (0.00%)
  Valori unici                   9,245  (31.93%)
  Valori float                   0

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            101.0000
  Max                            2,979,733
  Range                          2,979,632
  Media                          29,963.08
  Mediana                        1,528.00
  Moda

**Osservazioni:**
- **10.073 valori nulli** che corrispondono allo stesso numero di valori null della colonna `score`. Verifichiamo che si tratti delle stesse righe.
- Il dtype `float64`. Lo convertiamo a `Int64`.
- Gli outlier rilevati sono anime famosi con alta partecipazione, non errori nei dati.

In [53]:
# Verifica co-nullità
print(f'Null in score     : {df_details["score"].isna().sum():,}')
print(f'Null in scored_by : {df_details["scored_by"].isna().sum():,}')
print(f'Co-nulli (entrambi NaN) : {(df_details["score"].isna() & df_details["scored_by"].isna()).sum():,}')
print()

#Convertiamo a int
df_details['scored_by'] = df_details['scored_by'].astype('Int64')
print(f'scored_by dtype : {df_details["scored_by"].dtype}')

Null in score     : 10,073
Null in scored_by : 10,073
Co-nulli (entrambi NaN) : 10,073

scored_by dtype : Int64


Si verifica che i valori null su `scored_by` corrispondono ai valori null su `scored`.

**Nessuna pulizia necessaria**.

### 2.10 `start_date`

Data di inizio trasmissione, memorizzata come stringa. Usiamo `analyze` per capire il formato e valutare la conversione in un tipo datetime. I null sono attesi per gli anime `Not yet aired` senza data confermata.

In [54]:
df_details['start_date'] = df_details['start_date'].str.strip()
analyze(df_details['start_date'])


  Nome serie                     start_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,104  (97.06%)
  Null / NaN                     851  (2.94%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   9,367  (32.35%)
  Valori duplicati               18,737  (64.71%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori d

**Osservazioni:**
- **851 valori nulli** (~2.9%): strutturali, anime `Not yet aired` senza data confermata.
- **Lunghezza costante di 25 caratteri** e tutti gli esempi mostrano il pattern `YYYY-MM-DDTHH:MM:SS+00:00` con offset `+00:00` che rapresenta il timezone.
- **Conversione necessaria**: da stringa a `datetime64[us, UTC]` per abilitare operazioni temporali (confronti, differenze, etc). Usiamo UTC perché è la scelta coerente con l'offset `+00:00` già presente nei dati.
**Nessuna ulteriore pulizia necessaria.**

In [55]:
df_details['start_date'] = pd.to_datetime(df_details['start_date'], errors='coerce', utc=True)
print(f'start_date dtype: {df_details["start_date"].dtype}')

start_date dtype: datetime64[us, UTC]


### 2.11 `end_date`

Data di fine trasmissione, stessa struttura di `start_date`. I null sono più numerosi: gli anime `Currently Airing` e `Not yet aired` non hanno ancora una data di fine.

In [56]:
df_details['end_date'] = df_details['end_date'].str.strip()
analyze(df_details['end_date'])


  Nome serie                     end_date
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               11,267  (38.91%)
  Null / NaN                     17,688  (61.09%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   5,607  (19.36%)
  Valori duplicati               5,660  (19.55%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  25  caratteri
  Lunghezza max                  25  caratteri
  Lunghezza media                25.00  caratteri
  Lunghezza mediana              25.0000  caratteri
  Dev. standard lunghezza        0.00
  IQR lunghezza                  0.0000

Valori 

**Osservazioni:**
- **17.688 valori nulli** (~61.1%): strutturali, anime ancora in corso o non ancora iniziati.
- Il format è lo stesso di `start_date`. Facciamo la conversione a `datetime64[us, UTC]`.

In [57]:
df_details['end_date'] = pd.to_datetime(df_details['end_date'], errors='coerce', utc=True)
print(f'end_date dtype: {df_details["end_date"].dtype}')

end_date dtype: datetime64[us, UTC]


**Verifica consistenza `start_date` ↔ `end_date`**

Una volta convertite entrambe le date, verifichiamo che non ci siano righe in cui `end_date < start_date`: un anime non può terminare prima di iniziare. Consideriamo solo le righe in cui entrambe le date sono presenti, poiché confrontare con un valore nullo non ha significato.

In [58]:
# end_date < start_date
both_valid = df_details['start_date'].notna() & df_details['end_date'].notna()
inverted   = both_valid & (df_details['end_date'] < df_details['start_date'])
print(f'Righe con end_date < start_date: {inverted.sum()}')
print()
print(df_details[inverted][['mal_id', 'title', 'status', 'start_date', 'end_date']].head(10).to_string(index=False))

Righe con end_date < start_date: 101

 mal_id                                     title          status                start_date                  end_date
  48010                       23 Hao Niu Nai Tang Finished Airing 2020-03-01 00:00:00+00:00 2020-01-01 00:00:00+00:00
  23583                    Akage no Anne Specials Finished Airing 2001-02-11 00:00:00+00:00 2001-01-01 00:00:00+00:00
  44731                    Ali Diu Dongxi de Wawa Finished Airing 2012-08-31 00:00:00+00:00 2012-01-01 00:00:00+00:00
  53727                         Ashuai 8th Season Finished Airing 2022-06-03 00:00:00+00:00 2022-01-01 00:00:00+00:00
  45385                       AU: Kaixin Tongnian Finished Airing 2013-03-01 00:00:00+00:00 2013-01-01 00:00:00+00:00
  10330 Bakugan Battle Brawlers: Mechtanium Surge Finished Airing 2018-09-01 00:00:00+00:00 2018-01-01 00:00:00+00:00
  44551                       Baozou Kongbu Gushi Finished Airing 2014-07-11 00:00:00+00:00 2014-01-01 00:00:00+00:00
  44555           

**Osservazioni:**
- **101 righe** presentano `end_date < start_date`. Guardando i dati, il pattern è sempre lo stesso: `end_date` è il 1° gennaio di un anno (es. `2020-01-01`), mentre `start_date` è una data più tarda nello stesso anno (es. `2020-03-01`).
- La spiegazione è che a volte quando si conosce solo l'**anno** di fine trasmissione, non il giorno esatto si registra automaticamente il 1° gennaio come data di fine. Il risultato è che la data di fine appare prima della data di inizio, ma non è un errore reale, è solo una mancanza di precisione nel dato originale. Facciamo una ulteriore verifica per vedere se ci sono casi che rispettano questo pattern.

In [59]:
# Righe con end_date < start_date e end_date != 1° gennaio
both_valid = df_details['start_date'].notna() & df_details['end_date'].notna()
inverted   = both_valid & (df_details['end_date'] < df_details['start_date'])
not_jan1   = inverted & ~((df_details['end_date'].dt.month == 1) & (df_details['end_date'].dt.day == 1))
print(f'Righe con end_date < start_date e end_date != 1° gennaio: {not_jan1.sum()}')
print()
print(df_details[not_jan1][['mal_id', 'title', 'status', 'start_date', 'end_date']].to_string(index=False))

Righe con end_date < start_date e end_date != 1° gennaio: 8

 mal_id                                         title          status                start_date                  end_date
   3705                                Flashback Game Finished Airing 2001-10-18 00:00:00+00:00 2001-04-18 00:00:00+00:00
  10995                          Ganbare!! Nattou-san Finished Airing 2011-07-02 00:00:00+00:00 2011-07-01 00:00:00+00:00
  56892                     Kkomimanyeo Lara Season 2 Finished Airing 2023-10-14 00:00:00+00:00 2023-01-06 00:00:00+00:00
  34234                          Ohayou! Kokekkou-san Finished Airing 2016-10-02 00:00:00+00:00 2016-09-17 00:00:00+00:00
  43976                    Qi Jiguang Yingxiong Chuan Finished Airing 2012-06-01 00:00:00+00:00 2012-05-01 00:00:00+00:00
  38571                           Tobidase! Dokan-kun Finished Airing 2011-10-26 00:00:00+00:00 2011-03-14 00:00:00+00:00
  62453 Tobot: Daedosiui Yeongungdeul Season 2 Part 2 Finished Airing 2024-07-05 00:0

Abbiamo effettuato una ricerca e risulta che per questi anime non si conosce la data esatta di fine trasmissione. Siccome si conosce solo l'anno, li normaliziamo per corrispondere al resto dove la data di fine è impostata a 01 Gennaio dello stesso anno. C'è un ecezzione dove l'anime `Tobot: Daedosiui Yeongungdeul Season 2 Part 2` sembra essere terminato 1 anno prima della data di rilascio. Abbiamo verificato che si tratta di un errore di battitura. Anche questo è terminato nello stesso anno.

In [60]:
both_valid = df_details['start_date'].notna() & df_details['end_date'].notna()
inverted   = both_valid & (df_details['end_date'] < df_details['start_date'])
not_jan1   = inverted & ~((df_details['end_date'].dt.month == 1) & (df_details['end_date'].dt.day == 1))

df_details.loc[not_jan1, 'end_date'] = df_details.loc[not_jan1, 'start_date'].apply(
    lambda dt: dt.replace(month=1, day=1)
)

print(f'end_date corrette: {not_jan1.sum()}')
print()
print(df_details.loc[not_jan1.index[not_jan1], ['mal_id', 'title', 'start_date', 'end_date']].to_string(index=False))

end_date corrette: 8

 mal_id                                         title                start_date                  end_date
   3705                                Flashback Game 2001-10-18 00:00:00+00:00 2001-01-01 00:00:00+00:00
  10995                          Ganbare!! Nattou-san 2011-07-02 00:00:00+00:00 2011-01-01 00:00:00+00:00
  56892                     Kkomimanyeo Lara Season 2 2023-10-14 00:00:00+00:00 2023-01-01 00:00:00+00:00
  34234                          Ohayou! Kokekkou-san 2016-10-02 00:00:00+00:00 2016-01-01 00:00:00+00:00
  43976                    Qi Jiguang Yingxiong Chuan 2012-06-01 00:00:00+00:00 2012-01-01 00:00:00+00:00
  38571                           Tobidase! Dokan-kun 2011-10-26 00:00:00+00:00 2011-01-01 00:00:00+00:00
  62453 Tobot: Daedosiui Yeongungdeul Season 2 Part 2 2024-07-05 00:00:00+00:00 2024-01-01 00:00:00+00:00
  61524                      Xiongmao He Gan Mi Xiong 2025-05-04 00:00:00+00:00 2025-01-01 00:00:00+00:00


### 2.12 `synopsis`

Testo della sinossi dell'anime. I null sono attesi.

In [61]:
df_details['synopsis'] = df_details['synopsis'].str.strip()
analyze(df_details['synopsis'])


  Nome serie                     synopsis
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               23,828  (82.29%)
  Null / NaN                     5,127  (17.71%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   23,540  (81.30%)
  Valori duplicati               288  (0.99%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  10  caratteri
  Lunghezza max                  3,750  caratteri
  Lunghezza media                377.40  caratteri
  Lunghezza mediana              251.0000  caratteri
  Dev. standard lunghezza        351.38
  IQR lunghezza                  499.0000

V

**Osservazioni:**
- **5.127 valori nulli** (~17.7%): strutturali, potrebbe trattarsi di cortometraggi, CM (Commercial Message) e PV (Promotional Video) che spesso non hanno una sinossi registrata.
- **25 righe** contengono testo boilerplate di sistema (`"No synopsis has been added for this series yet."`): da convertire a `null`.
- **288 valori duplicati**: da investigare, potrebbero nascondere ulteriori pattern boilerplate.
- **3 sinossi contengono URL** e **38 contengono `@`**: da investigare per verificare che si tratti di contenuto legittimo.

**Duplicati nella sinossi:**

`analyze()` riporta 288 valori duplicati. Esaminiamo i testi più ricorrenti per verificare se esistono altri pattern boilerplate non catturati dal filtro principale, oppure se si tratta di duplicati legittimi (sinossi condivise tra stagioni o titoli correlati).

In [62]:
# Duplicati in synopsis (su dati non ancora puliti)
dup_mask_syn = df_details['synopsis'].duplicated(keep=False) & df_details['synopsis'].notna()
dup_counts = df_details.loc[dup_mask_syn, 'synopsis'].value_counts()

print(f'Righe con synopsis duplicata: {dup_mask_syn.sum()}')
print(f'Testi univoci tra i duplicati: {len(dup_counts)}')
print()

top15 = dup_counts.head(15).reset_index()
top15.columns = ['synopsis', 'count']
top15['preview'] = top15['synopsis'].str[:120].str.replace('\n', ' ', regex=False).apply(repr)
top15['line'] = top15['count'].astype(str).str.rjust(3) + 'x  ' + top15['preview']

print('Top 15 synopsis duplicate (prime 120 caratteri):')
print('─' * 80)
print('\n'.join('  ' + top15['line']))

Righe con synopsis duplicata: 456
Testi univoci tra i duplicati: 168

Top 15 synopsis duplicate (prime 120 caratteri):
────────────────────────────────────────────────────────────────────────────────
   23x  'No synopsis has been added for this series yet.  Click here to update this information.'
   12x  'Furukawa Taku film.'
   12x  'Film by Takashi Ito.'
   10x  'A short animation by Taku Furukawa.'
    7x  'A short film by Okamoto Tadanari.'
    6x  'Short animation by Rapparu.'
    6x  'Short film by Hirano Ryou.'
    6x  'A short puppet animation movie by Tadahito Mochinaga.'
    6x  'Short film by Kurosaka Keita.'
    5x  'A short film by Haruka Suzuki.'
    5x  'Unaired specials included in the Blu-ray and DVD release.'
    4x  'A mix of animation and live-action.'
    4x  'Three episodes aired on TV as part of the Ribon x Oha Suta specials.'
    4x  "The music film is a part of TOHO animation's 10th anniversary project."
    4x  'Short animation by Komazaki Tomomi.'


**Risultati indagine duplicati:**

- **456 righe coinvolte** in duplicazioni (168 testi univoci distinti). Il numero sembra diverso dal numero 288 riportato da `analyze()` perché quella cifra conta solo le copie non-prime, mentre qui si contano tutte le righe con testo non unico.
- Il testo più frequente (23×) è il boilerplate `"No synopsis has been added for this series yet..."` che rimuoviamo successivamente.
- Tutti gli altri duplicati sono descrizioni brevi e generiche di cortometraggi d'autore o descrizioni standard di release. Si tratta di contenuto corretto.
- **Nessun nuovo pattern boilerplate** da rimuovere. I duplicati rimanenti sono legittimi.

**Sinossi contenenti URL e `@`:**

`analyze()` riporta 3 sinossi con URL e 38 con `@`. Verifichiamo il contesto per determinare se si tratta di contenuto legittimo o artefatti da rimuovere.

In [64]:
import re

# Sinossi contenenti URL
url_mask_syn = df_details['synopsis'].str.contains(r'https?://\S+', regex=True, na=False)
print(f'Sinossi con URL: {url_mask_syn.sum()}')
print()

df_url = df_details.loc[url_mask_syn, ['mal_id', 'title', 'type']].copy()
df_url['synopsis'] = df_details.loc[url_mask_syn, 'synopsis'].str.findall(r'https?://\S+')
df_url = df_url.explode('synopsis').reset_index(drop=True)
print(df_url.to_string(index=False))

print()
print('─' * 80)
print()

# Sinossi contenenti '@'
at_mask_syn = df_details['synopsis'].str.contains(r'@', regex=False, na=False)
print(f'Sinossi con @: {at_mask_syn.sum()}')
print()

df_at = df_details.loc[at_mask_syn, ['mal_id', 'title']].copy()
df_at['synopsis'] = df_details.loc[at_mask_syn, 'synopsis'].str.findall(r'\S*@\S+')
df_at = df_at.explode('synopsis').reset_index(drop=True)
print(df_at.to_string(index=False))

Sinossi con URL: 3

 mal_id                       title    type                                       synopsis
  47634 Kamen Senshi Lavithunder CM Special                  https://vgmdb.net/album/54883
   6230        Kyokujitsu no Kantai     OVA http://en.wikipedia.org/wiki/Konpeki_no_Kantai
  43196              Luotuo Xian Wu   Movie               https://jouz.myblog.arts.ac.uk/,

────────────────────────────────────────────────────────────────────────────────

Sinossi con @: 38

 mal_id                                                                                title                 synopsis
  35651                                                                  Aozora Under Girls!               iDOLM@STER
  48814                                                                                 Blue                     m@e.
  34489                                Cinderella Girls Gekijou: 5-shuunen Kinen Short Anime               iDOLM@STER
  35346                                   

- Tutte le sinossi contenenti URL sono citazioni di fonte.
- In tutti i casi contenenti `@`, si tratta di un carattere intenzionale di titoli o nomi ufficiali.
- Rimuoviamo solo i testi bolerplate impostandoli a null.

**Nessuna ulteriore pulizia necessaria.**

In [63]:
pattern = r'^No synopsis has been added for this series yet'
mask_bp = df_details['synopsis'].str.contains(pattern, regex=True, na=False)
print(f'Righe con sinossi boilerplate: {mask_bp.sum()}')
print(df_details.loc[mask_bp, 'synopsis'].value_counts().to_string())

df_details.loc[mask_bp, 'synopsis'] = np.nan
print(f'\nNull in synopsis dopo pulizia: {df_details["synopsis"].isna().sum():,}')

Righe con sinossi boilerplate: 25
synopsis
No synopsis has been added for this series yet.\n\nClick here to update this information.     23
No synopsis has been added for this series yet.                                                1
No synopsis has been added for this series yet. \n\nClick here to update this information.     1

Null in synopsis dopo pulizia: 5,152


### 2.13 `rank`

Classifica globale basata sullo score. I null sono attesi in quanto anime senza score non ricevono un rank.

In [ ]:
analyze(df_details['rank'])

**Osservazioni:**
- **6.958 valori nulli** (~24.0%): strutturali, anime non classificati (score assente).
- **17.765 valori unici su 21.997 non-null**: 4.232 anime condividono il rank con almeno un altro quando più anime hanno lo stesso score.
- **Conversione dtype**: da `float64` a `Int64` dato che i rank sono interi.
- Si nota la presenza di bucchi e rank condivisi (Max = 22,020 mentre il totale di valori non nulli è 21,977). Questo è dovuto al fatto che `rank` è un attributo calcolato da MAL, dove se più anime condividono lo stesso score, ricevono lo stesso rank. Non si tratta di errori nei dati e modificare questi valori produrrebbe una colonna diversa da quella originale.

**Nessuna ulteriore pulizia necessaria.**

In [ ]:
df_details['rank'] = df_details['rank'].astype('Int64')
print(f'rank dtype: {df_details["rank"].dtype}')

### 2.14 `popularity`

Classifica di popolarità basata sul numero di utenti che hanno l'anime nella propria lista.

In [ ]:
analyze(df_details['popularity'])

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un valore.
- **Dtype corretto**: `int64`.
- Si nota la presenza di bucchi e popolarità condivise (Max = 28,998 mentre il totale di valori non nulli è 28,955). Questo è dovuto al fatto che `popularity` è un attributo calcolato da MAL, dove se più anime condividono lo stesso numero di `members`, ricevono la stessa popolarità. Non si tratta di errori nei dati e modificare questi valori produrrebbe una colonna diversa da quella originale.

**Nessuna pulizia necessaria.**

### 2.15 `members`

Numero di utenti MAL che hanno l'anime nella propria lista.

In [ ]:
analyze(df_details['members'])

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un valore.
- **Dtype corretto**: `int64`.

**Nessuna pulizia necessaria**

### 2.16 `favorites`

Numero di utenti che hanno aggiunto l'anime ai propri preferiti.

In [ ]:
analyze(df_details['favorites'])

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un valore.
- **Dtype corretto**: `int64`.
- Il 44.6% dei valori è 0. Non è un anomalia ma si tratta di anime poco conosciuti.

Nessuna pulizia necessaria.